core/rawdb/accessors_chain.go

```diff
@@ -815,129 +816,18 @@ func WriteAncientBlocks(db ethdb.AncientWriter, blocks []*types.Block, receipts
 			if i > 0 {
 				tdSum.Add(tdSum, header.Difficulty)
 			}
-			if err := writeAncientBlock(op, block, header, stReceipts, tdSum); err != nil {
-				return err
-			}
-		}
-		return nil
-	})
-}
-
-// WriteAncientBlocksAfterCancun writes entire block data into ancient store and returns the total written size.
-func WriteAncientBlocksAfterCancun(db ethdb.AncientStore, config *params.ChainConfig, blocks []*types.Block, receipts []types.Receipts, td *big.Int) (int64, error) {
-	if len(blocks) == 0 {
-		return 0, nil
-	}
-
-	cancunIndex := -1
-	for i, block := range blocks {
-		if config.IsCancun(block.Number(), block.Time()) {
-			cancunIndex = i
-			break
-		}
-	}
-	log.Info("WriteAncientBlocksAfterCancun", "startAt", blocks[0].Number(), "cancunIndex", cancunIndex, "len", len(blocks))
-	if cancunIndex < 0 {
-		return WriteAncientBlocks(db, blocks, receipts, td)
-	}
-
-	// if there has pre-cancun and post-cancun blocks, write them separately
-	var (
-		tdSum      = new(big.Int).Set(td)
-		stReceipts []*types.ReceiptForStorage
-	)
-
-	// write pre-cancun blocks
-	preBlocks := blocks[:cancunIndex]
-	preReceipts := receipts[:cancunIndex]
-	preSize, err := db.ModifyAncients(func(op ethdb.AncientWriteOp) error {
-		for i, block := range preBlocks {
-			// Convert receipts to storage format and sum up total difficulty.
-			stReceipts = stReceipts[:0]
-			for _, receipt := range preReceipts[i] {
-				stReceipts = append(stReceipts, (*types.ReceiptForStorage)(receipt))
-			}
-			header := block.Header()
-			if i > 0 {
-				tdSum.Add(tdSum, header.Difficulty)
+			if !blobSidecarFound && len(block.Sidecars()) > 0 {
```

Comment:
The optmize is great, may reduce some duplicated code. But I think we also need use `config.IsCancun` not just check `len(block.Sidecars()) > 0`, because the freezerDB need to save items one by one, and it is possible the block contains no blob after cancun.

Reply:
Here is a another issue that we cannot exe `ResetEmptyBlobAncientTable` inner `WriteAncientBlocksWithSidecars`, because it will remove useless file and renew table & bacher.

Reply:
How about we just `ResetEmptyBlobAncientTable` and save blobs ancient if we find the last block enable hard fork? It will be simpler too, and do not break above rules. And it's ok to save empty blob for some pre-cancun blocks too.

```go
writeBlob := false
if config.IsCancun(last.number, last.time) {
    writeBlob = true
   ResetEmptyBlobAncientTable(first.number)
}
WriteAncientBlocks(db, chain, receipts, td, writeBlob)
```